presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java

```diff
@@ -81,24 +87,74 @@ public static Slice charToVarcharCast(@LiteralParameter("x") Long x, @LiteralPar
     @ScalarOperator(OperatorType.SATURATED_FLOOR_CAST)
     @SqlType("char(y)")
     @LiteralParameters({"x", "y"})
-    // This function returns Char(y) value that is smaller than original Varchar(x) value. However, it is not necessarily the largest
-    // Char(y) value that is smaller than the original Varchar(x) value. This is fine though for usage in TupleDomainTranslator.
-    public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") Long y, @SqlType("varchar(x)") Slice slice)
+    public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long y, @SqlType("varchar(x)") Slice slice)
     {
-        Slice trimmedSlice = trimSpaces(slice);
-        int trimmedTextLength = countCodePoints(trimmedSlice);
-        int numberOfTrailingSpaces = slice.length() - trimmedSlice.length();
+        List<Integer> codePoints = new ArrayList<>(toCodePoints(slice));
```

Comment:
Fair question. I'm not fluent in Slice API and related classes, so maybe there is a way, but I could not find any. Previously the code operated on Slices, but I had two problems with that:

- decrementing last code point -- I don't know how to do it with slices
- appending MAX_CODE_POINTs up to char length -- in Slice I cannot append, so I'd need to concatenate two Slices (how?)
- looking at n'th char in Slice is linear operation since chars are of varying byte length

Also it seems materializing a list is "only" readability problem, not performance, because saturated floor cast is used during planning, not at query execution time.